examples/src/main/java/alluxio/cli/UnderFileSystemContractTest.java

```diff
@@ -0,0 +1,325 @@
+    mConf.set(PropertyKey.UNDERFS_S3A_STREAMING_UPLOAD_PARTITION_SIZE, "5MB");
+    mConf.set(PropertyKey.UNDERFS_S3A_INTERMEDIATE_UPLOAD_CLEAN_AGE, "0");
+
+    createUnderFileSystem();
+    String testDir = PathUtils.concatPath(mUfsPath, UUID.randomUUID());
+    loadAndRunTests(new S3ASpecificOperations(testDir, mUfs, mConf), testDir);
+  }
+
+  /**
+   * Loads and runs the tests.
+   *
+   * @param operationsToTest the class that contains the tests to run
+   * @param testDir the test directory to run tests against
+   */
+  private void loadAndRunTests(Object operationsToTest, String testDir) throws Exception {
+    try {
+      Class classToRun = operationsToTest.getClass();
+      Field[] fields = classToRun.getDeclaredFields();
+      for (Field field : fields) {
+        field.setAccessible(true);
+      }
+      Method[] tests = classToRun.getDeclaredMethods();
+      for (Method test : tests) {
+        String testName = test.getName();
+        if (testName.endsWith("Test")) {
+          LOG.info("Running test: " + testName);
+          try {
+            test.invoke(operationsToTest);
+          } catch (InvocationTargetException e) {
+            if (mIsS3Ufs) {
+              List<String> operations = getRelatedS3AOperations(testName);
+              if (operations.size() > 0) {
+                LOG.info("Related S3 operations: "
+                    + StringUtils.join(operations, ","));
+              }
+            }
+            throw new IOException(e.getTargetException());
+          }
+          LOG.info("Test Passed!");
+          cleanupUfs(testDir);
+        }
+      }
+    } finally {
+      mUfs.deleteDirectory(testDir, DeleteOptions.defaults().setRecursive(true));
+      mUfs.close();
+    }
+  }
+
+  /**
+   * Creates the under file system.
+   */
+  private void createUnderFileSystem() {
+    mUfs = mFactory.create(mUfsPath,
+        UnderFileSystemConfiguration.defaults(mConf));
+    if (mUfs == null) {
+      LOG.error("Failed to create under filesystem");
+      System.exit(1);
+    }
+  }
+
+  /**
+   * Cleans all the files or sub directories inside the given directory
+   * in the under filesystem.
+   *
+   * @param directory the directory to clean
+   */
+  private void cleanupUfs(String directory) throws IOException {
+    UfsStatus[] statuses = mUfs.listStatus(directory);
+    for (UfsStatus status : statuses) {
+      if (status instanceof UfsFileStatus) {
+        mUfs.deleteFile(PathUtils.concatPath(directory, status.getName()));
+      } else {
+        mUfs.deleteDirectory(PathUtils.concatPath(directory, status.getName()),
+            DeleteOptions.defaults().setRecursive(true));
+      }
+    }
+  }
+
+  /**
+   * Gets the S3A operations related to the failed test. This method
+   * should only be called when the ufs is S3A.
+   *
+   * @param testName the name of the failed test
+   * @return the related S3A operations
+   */
+  private List<String> getRelatedS3AOperations(String testName) {
```

Comment:
will using annotations on methods in `S3ASpecificOperations` and using reflection to query annotations make this cleaner? not sure if there's a way around this ugly `switch`